include/pisa/index.hpp

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include <any>
+#include <optional>
+
+#include <gsl/span>
+
+#include "index_metadata.hpp"
+
+namespace pisa {
+
+using DocId = std::uint32_t;
+using TermId = std::uint32_t;
+
+using OffsetSpan = gsl::span<std::size_t const>;
+using BinarySpan = gsl::span<std::byte const>;
+
+template <typename Cursor>
+class Reader;
+
+/// Note that offsets is stored here uncompressed.
+/// For compressed, this needs to be slightly modified but the general idea is the same.
+/// We'll need something that can get offsets for posting lists.
+struct PostingData {
+    BinarySpan postings;
+    OffsetSpan offsets;
+};
+
+class Index {
+   public:
+    template <typename Source>
+    Index(gsl::span<std::uint32_t const> document_lengths, float avg_document_length, Source source)
+        : m_document_lengths(document_lengths),
+          m_avg_document_length(avg_document_length),
+          m_source(std::move(source))
+    {
+    }
+    [[nodiscard]] auto num_terms() const -> std::size_t;
+    [[nodiscard]] auto num_documents() const -> std::size_t;
+    [[nodiscard]] auto document_length(DocId docid) const -> std::uint32_t;
+    [[nodiscard]] auto avg_document_length() const -> float;
+    [[nodiscard]] auto normalized_document_length(DocId docid) const -> float;
+
+   protected:
+    [[nodiscard]] auto fetch_postings(TermId term, PostingData const& data) const
+        -> gsl::span<std::byte const>;
+
+   private:
+    gsl::span<std::uint32_t const> m_document_lengths;
```

Comment:
Let's think is we really want to store the document lengths. True that we need those for scoring, but do we assume that we will always score? 
I believe this data should be part of some scoring metadata file.